internal/engine/upper_test.go

```diff
@@ -0,0 +1,143 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/windmilleng/fsnotify"
+	"github.com/windmilleng/tilt/internal/logger"
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/watch"
+)
+
+type fakeBuildAndDeployer struct {
+	startedServices []model.Service
+	cond            *sync.Cond
+}
+
+var _ BuildAndDeployer = &fakeBuildAndDeployer{}
+
+func (b *fakeBuildAndDeployer) BuildAndDeploy(ctx context.Context, service model.Service, token BuildToken) (BuildToken, error) {
+	b.startedServices = append(b.startedServices, service)
+	b.cond.L.Lock()
+	b.cond.Broadcast()
+	b.cond.L.Unlock()
+	return nil, nil
+}
+
+func newFakeBuildAndDeployer() *fakeBuildAndDeployer {
+	return &fakeBuildAndDeployer{cond: sync.NewCond(&sync.Mutex{})}
+}
+
+type fakeNotify struct {
+	paths  []string
+	events chan fsnotify.Event
+	errors chan error
+}
+
+func (n *fakeNotify) Add(name string) error {
+	n.paths = append(n.paths, name)
+	return nil
+}
+
+func (n *fakeNotify) Close() error {
+	return nil
+}
+
+func (n *fakeNotify) Errors() chan error {
+	return n.errors
+}
+
+func (n *fakeNotify) Events() chan fsnotify.Event {
+	return n.events
+}
+
+func newFakeNotify() *fakeNotify {
+	return &fakeNotify{paths: make([]string, 0), errors: make(chan error, 1), events: make(chan fsnotify.Event, 10)}
+}
+
+var _ watch.Notify = &fakeNotify{}
+
+func TestUpper_Up(t *testing.T) {
+	f := newTestFixture(t)
+	service := model.Service{Name: "foobar"}
+	err := f.upper.Up(f.context, service, false, os.Stdout, os.Stderr)
+	assert.Nil(t, err)
+	assert.Equal(t, []model.Service{service}, f.b.startedServices)
+}
+
+func TestUpper_UpWatchZeroRepos(t *testing.T) {
+	f := newTestFixture(t)
+	service := model.Service{Name: "foobar"}
+	err := f.upper.Up(f.context, service, true, os.Stdout, os.Stderr)
+	if assert.NotNil(t, err) {
+		assert.Contains(t, err.Error(), "0 repos")
+	}
+}
+
+func TestUpper_UpWatchError(t *testing.T) {
+	f := newTestFixture(t)
+	mount := model.Mount{Repo: model.LocalGithubRepo{LocalPath: "/go"}, ContainerPath: "/go"}
+	service := model.Service{Name: "foobar", Mounts: []model.Mount{mount}}
+	f.b.cond.L.Lock()
```

Comment:
let me come talk to you about these condition variables, i think you're confused about something

Reply:
yeah, buffered channels are much prettier